### ISSUE TITLE
proposal: enable some extra default features in `--dev` mode (with an opt-out?)

### ISSUE THREAD

**Issue by EverlastingBugstopper** [apollographql/router#2125]
## Context

We've received numerous requests recently for a few ergonomics improvements to `rover dev`, most notably CORS and header propagation.

I'd really like to avoid exposing an entire config file via `rover dev` right now, especially given our possible plans for cloudy rover dev down the line where the config file will be pulled from a central location.

However, there is a real need for these settings to be applied. 

### Related issues

- https://github.com/apollographql/rover/issues/1373
- https://github.com/apollographql/rover/issues/1345

## Proposal

When `--dev` is passed to the router, do the following:

1) Enable most permissive CORS settings (or perhaps just open up requests to all localhost origins?) by default
2) Provide a flag to disable this new permissive CORS default (`--no-permissive-cors` or something?) we can expose this in `rover dev` as well (and it should be applicable to both unauthenticated dev sessions and possible future authenticated dev sessions)
3) Pass through all headers to the subgraphs by default
4) Provide a flag to disable the passthrough headers (`--no-passthrough-headers`?) again, we can expose that in `rover dev`.

**Comment by o0Ignition0o** [comment]
Two cents: I think cors mirror in rover dev could make sense but I fear ppl would be too surprised with things not working in prod, it might work if we were super vocal with a log message like “we went with super loose cors settings, this might end poorly if you deploy, run without dev before you do” or something ?

**Comment by EverlastingBugstopper** [comment]
That makes sense to me! It may make sense to research what other dev tools do here as well. For instance, what does lambda do if you run things locally vs. in prod?

**Comment by abernix** [comment]
> Enable most permissive CORS settings (or perhaps just open up requests to all localhost origins?) by default


I'm interested in exploring what other dev tools do, but as I noted in #1898, it makes me uncomfortable to facilitate the most permissive settings in dev mode, or even on `localhost`.  Those can _still_ be used as vectors, and the user _will_ need to configure something to be successful in production if they are not running on the same domain.

**Comment by EverlastingBugstopper** [comment]
@abernix that makes sense to me but it also makes me sad.

OK. In that case we need to provide some other opt-in way to configure this.

For CORS, I think opt-in usage allowing multiple CORS settings is a decent strategy?

```rust
use crate::Result;
use clap::Parser;
use std::collections::HashMap;

#[derive(Parser, Debug)
struct CorsOpts {
  #[arg(long)]
  cors: Vec<String>
}

impl CorsOpts {
  fn get_allowed_hosts(&mut self) -> Vec<String> {
	cors.push("https://studio.apollographql.com".to_string());
	cors
  }
}
```

Unfortunately this option would only be able to be set by the _first_ `rover dev` session, so we should either hard error or print a warning saying we're not using that option in this invocation since we're attaching to an existing session.

For headers... I'm at a loss. It seems like complex configuration that is hard to cleanly map to CLI arguments (but we could probably spend some time thinking about it). Perhaps the first step just to get _some_ folks unblocked is adding an opt-in `--passthrough-headers all` argument. can we think of other useful values to pass to that argument? it would default to `none`, are there other rules that could fit in here?

---

I'm really hesitant to allow a `--router-config` option to be passed to `rover dev` because of breaking changes made to the configuration down the line, but maybe it's something we just need to do. We could also do something a bit more interstitial (and maybe painful) where you spin up a `rover dev` session and then you run `EDITOR=vim rover dev --edit-config` and you can edit the config and write to it to reload the router with those options but.... that feels very not ideal.

**Comment by BrynCooke** [comment]
Honestly I think that rover dev needs first class router.yaml functionality baked in to it. Users do lots of interesting things with headers, not just propagation, and it's only going to get worse as we start enabling more router features for graphos.

**Comment by jpvajda** [comment]
I'd just thought I'd add: some of the DX ideas we've been discussing to enable client developers to be able to use Rover CLI in 2023 for getting started experiences building a simple full stack graphQL app would definitely require this capability.

**Comment by EverlastingBugstopper** [comment]
Closing as won't do, instead we will be adding `--router-config` to `rover dev` (tracked [here](https://github.com/apollographql/rover/pull/1446)).
